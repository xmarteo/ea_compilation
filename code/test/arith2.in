7
4
89
0
-2
-345
1
4
