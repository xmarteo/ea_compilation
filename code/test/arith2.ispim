Exception occurred at PC=0x00400268
  Unaligned address in inst/data fetch: 0x20040009
  Exception 4  [Address error in inst/data fetch]  occurred and ignored
Exception occurred at PC=0x00400290
  Unaligned address in store: 0x20040009
  Exception 5  [Address error in store]  occurred and ignored
Exception occurred at PC=0x00400268
  Unaligned address in inst/data fetch: 0x2004000d
  Exception 4  [Address error in inst/data fetch]  occurred and ignored
Exception occurred at PC=0x00400290
  Unaligned address in store: 0x2004000d
  Exception 5  [Address error in store]  occurred and ignored
Exception occurred at PC=0x00400268
  Unaligned address in inst/data fetch: 0x20040011
  Exception 4  [Address error in inst/data fetch]  occurred and ignored
Exception occurred at PC=0x00400290
  Unaligned address in store: 0x20040011
  Exception 5  [Address error in store]  occurred and ignored
Exception occurred at PC=0x00400268
  Unaligned address in inst/data fetch: 0x20040015
  Exception 4  [Address error in inst/data fetch]  occurred and ignored
Exception occurred at PC=0x00400290
  Unaligned address in store: 0x20040015
  Exception 5  [Address error in store]  occurred and ignored
Exception occurred at PC=0x00400268
  Unaligned address in inst/data fetch: 0x20040019
  Exception 4  [Address error in inst/data fetch]  occurred and ignored
Exception occurred at PC=0x00400290
  Unaligned address in store: 0x20040019
  Exception 5  [Address error in store]  occurred and ignored
Exception occurred at PC=0x00400268
  Unaligned address in inst/data fetch: 0x2004001d
  Exception 4  [Address error in inst/data fetch]  occurred and ignored
Exception occurred at PC=0x00400290
  Unaligned address in store: 0x2004001d
  Exception 5  [Address error in store]  occurred and ignored
Exception occurred at PC=0x00400268
  Unaligned address in inst/data fetch: 0x20040021
  Exception 4  [Address error in inst/data fetch]  occurred and ignored
Exception occurred at PC=0x00400290
  Unaligned address in store: 0x20040021
  Exception 5  [Address error in store]  occurred and ignored
4
89
0
-2
-345
1
4
