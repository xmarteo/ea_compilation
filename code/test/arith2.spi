.data                  
globals:               
.space 16           
heap_info:
	.space 12
    .text                 
main:                 
la    $gp, globals    
# allocate the heap   
li    $a0, 65536        
li    $v0, 9          
syscall               
# saving heap size    
move  $a2, $a0        
# computing heap_info 
la    $a0, heap_info  
# beginning of heap   
sw    $v0, 0($a0)     
# current position    
sw    $v0, 4($a0)     
# end of heap         
add   $a2, $a2, $v0   
sw    $a2, 8($a0)     
j     _main122              
write:                
li    $v0, 1          
syscall               
j     $ra             
.data                 
nl:                   
.asciiz "\n"       
.align 2              
.text                 
writeln:              
li    $v0, 1          
syscall               
la    $a0, nl         
li    $v0, 4          
syscall               
j     $ra             
readln:               
li    $v0, 5          
syscall               
j     $ra             
alloc:                
la    $a1, heap_info  
# adding extra word   
addi  $t1, $a0, 4     
# current position    
lw    $v0, 4($a1)     
move  $t0, $v0        
# end of the heap     
lw    $a2, 8($a1)     
add   $v0, $t1, $v0   
# must have v0<=a2    
bgt   $v0, $a2, oom   
sw    $v0, 4($a1)     
# length is written   
move  $v0, $t0        
div   $a0, $a0, 4     
sw    $a0, 0($v0)     
addi  $v0, $v0, 4     
j     $ra             
.data                 
error_str:            
.asciiz "out0fMem"  
.align 2              
.text                 
oom:                  
# means out of mem.   
la $a0, error_str     
li $v0, 4             
syscall               
li $v0, 10            
syscall               
j     $ra             

# begin double
double67:
li    $v0, 0
addiu $a1, $a0, -1
li    $a0, 2
div   $a0, $a1, $a0
bgtz  $a0, double49
addiu $a1, $v0, -1
li    $v0, 2
div   $v0, $a1, $v0
addu  $a0, $v0, $a0
subu  $v0, $v0, $a0
sll   $v0, $v0, 1
addiu $v0, $v0, 1
double69:
jr    $ra
double49:
bgtz  $zero, double69
addiu $a1, $v0, -1
li    $v0, 2
div   $v0, $a1, $v0
subu  $a1, $v0, $a0
sll   $a1, $a1, 1
addiu $a1, $a1, 1
addu  $v0, $a0, $v0
addiu $a2, $a1, -1
li    $a1, 2
div   $a1, $a2, $a1
addu  $v0, $v0, $a1
sll   $v0, $v0, 1
addiu $v0, $v0, 1
addiu $a1, $v0, -1
li    $v0, 2
div   $a1, $a1, $v0
li    $v0, -2
subu  $a0, $zero, $a0
mul   $v0, $v0, $a0
addu  $v0, $a1, $v0
sll   $v0, $v0, 1
addiu $v0, $v0, 1
j     double69
# end double

# begin arraydouble
arraydouble68:
addiu $sp, $sp, -20
sw    $ra, 16($sp)
sw    $s3, 12($sp)
sw    $s2, 8($sp)
sw    $s1, 4($sp)
sw    $s0, 0($sp)
move  $s2, $a0
move  $s3, $a1
li    $s1, 0
arraydouble55:
addiu $a0, $s1, -1
li    $v0, 2
div   $v0, $a0, $v0
addiu $a1, $s3, -1
li    $a0, 2
div   $a0, $a1, $a0
slt   $a1, $v0, $a0
bgtz  $a1, arraydouble46
lw    $ra, 16($sp)
lw    $s3, 12($sp)
lw    $s2, 8($sp)
lw    $s1, 4($sp)
lw    $s0, 0($sp)
addiu $sp, $sp, 20
jr    $ra
arraydouble46:
bgtz  $zero, arraydouble42
sll   $a0, $v0, 3
li    $a1, -8
mul   $a1, $a1, $v0
addu  $a1, $a1, $v0
addu  $a0, $a0, $a1
sll   $a0, $a0, 2
addu  $s0, $s2, $a0
sll   $a0, $v0, 1
subu  $v0, $a0, $v0
sll   $v0, $v0, 2
addu  $v0, $s2, $v0
lw    $v0, 0($v0)
sll   $v0, $v0, 1
addiu $a0, $v0, 1
jal   double67
addiu $a0, $v0, -1
li    $v0, 2
div   $v0, $a0, $v0
sw    $v0, 0($s0)
addiu $a0, $s1, -1
li    $v0, 2
div   $v0, $a0, $v0
sll   $v0, $v0, 1
addiu $s1, $v0, 3
j     arraydouble55
arraydouble42:
sll   $v0, $a0, 1
addiu $s1, $v0, 1
j     arraydouble55
# end arraydouble

# begin _main
_main122:
addiu $sp, $sp, -4
sw    $ra, 0($sp)
jal   readln
sll   $v0, $v0, 1
addiu $v0, $v0, 1
sw    $v0, 12($gp)
lw    $v0, 12($gp)
addiu $a0, $v0, -1
li    $v0, 2
div   $v0, $a0, $v0
subu  $a0, $zero, $v0
li    $v0, -1
div   $v0, $a0, $v0
sll   $a0, $v0, 2
jal   alloc
sw    $v0, 0($gp)
lw    $v0, 12($gp)
addiu $a0, $v0, -1
li    $v0, 2
div   $v0, $a0, $v0
sll   $v0, $v0, 1
addiu $v0, $v0, 1
sw    $v0, 8($gp)
_main90:
lw    $v0, 12($gp)
addiu $a0, $v0, -1
li    $v0, 2
div   $v0, $a0, $v0
bgtz  $v0, _main74
bne   $v0, $zero, _main75
_main50:
lw    $v0, 0($gp)
sll   $v0, $v0, 1
addiu $a0, $v0, 1
lw    $v0, 8($gp)
addiu $a1, $v0, -1
li    $v0, 2
div   $v0, $a1, $v0
sll   $v0, $v0, 1
addiu $a1, $v0, 1
jal   arraydouble68
lw    $v0, 8($gp)
addiu $a0, $v0, -1
li    $v0, 2
div   $v0, $a0, $v0
sll   $v0, $v0, 1
addiu $v0, $v0, 1
sw    $v0, 12($gp)
_main33:
lw    $v0, 12($gp)
addiu $a0, $v0, -1
li    $v0, 2
div   $v0, $a0, $v0
beq   $v0, $zero, _main133
bgtz  $v0, _main17
bgez  $v0, _main17
_main133:
lw    $ra, 0($sp)
addiu $sp, $sp, 4
jr    $ra
_main17:
lw    $v0, 12($gp)
addiu $a0, $v0, -1
li    $v0, 2
div   $v0, $a0, $v0
sll   $v0, $v0, 1
addiu $v0, $v0, -1
sw    $v0, 12($gp)
lw    $a1, 0($gp)
lw    $v0, 12($gp)
addiu $a0, $v0, -1
li    $v0, 2
div   $v0, $a0, $v0
sll   $v0, $v0, 2
addu  $v0, $a1, $v0
lw    $a0, 0($v0)
jal   writeln
j     _main33
_main75:
bltz  $v0, _main50
_main74:
lw    $v0, 12($gp)
addiu $a0, $v0, -1
li    $v0, 2
div   $v0, $a0, $v0
sll   $v0, $v0, 1
addiu $v0, $v0, -1
sw    $v0, 12($gp)
jal   readln
sll   $v0, $v0, 1
addiu $v0, $v0, 1
sw    $v0, 4($gp)
lw    $a1, 0($gp)
lw    $v0, 12($gp)
addiu $a0, $v0, -1
li    $v0, 2
div   $v0, $a0, $v0
sll   $v0, $v0, 2
addu  $a1, $a1, $v0
lw    $v0, 4($gp)
addiu $a0, $v0, -1
li    $v0, 2
div   $v0, $a0, $v0
sw    $v0, 0($a1)
j     _main90
# end _main
