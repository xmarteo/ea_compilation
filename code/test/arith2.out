8
178
0
2
345
2
8
